src/common/melody_generator.py

```diff
@@ -0,0 +1,134 @@
+
+            # inference stage, this will result in a distribution of probabilities, but we just want the first/most likely one
+            # i.e: [0.]
+            probabilities = self.model.predict(onehot_seed)[0]
+            output_int = self._sample_with_temperature(probabilities, temperature)
+
+            # update seed
+            seed.append(output_int)
+
+            # map int to our encoding
+            output_symbol = [k for k, v in self._mappings.items() if v == output_int][0]
+
+            # check whether we're at the end of a melody
+            if output_symbol == "/":
+                break
+
+            # update the melody
+            melody.append(output_symbol)
+
+        return melody
+
+
+    def _sample_with_temperature(self, probabilities, temperature=0.7):
+        """ 
+        The temperature modifies the probability distribution
+        the lower the temperature, the more deterministic the sampling is
+        if the temperature is 1 that is just default settings
+        higher temperature, more exploritive model
+        """
+        predicitions = np.log(probabilities) / temperature
+        probabilities =  np.exp(predicitions) / np.sum(np.exp(predicitions))
+
+        choices = range(len(probabilities))
+        index = np.random.choice(choices, p=probabilities)
+
+        return index
+
+    def save_melody(self, melody, file_name, format="midi", step_duration=0.25):
+        """ 
+            * The melody object you want to pass
+            * the name of the saved file
+            * the file format you want to save it in
+            * The step duration you trained the model in, 0.25 is a sixteenth note
+        """
+
+        # create a music21 stream
+        stream = m21.stream.Stream()
+
+        # parse all the symbols in the melody and create note/rest objects
+        # i.e: 60 _ _ _ r _ 62 _ ....
+        start_symbol = None
+        step_counter = 1
+
+        for i, symbol in enumerate(melody):
+
+            # handle case in which we have a note/rest
+            # else handle case in which we have a prolongation sign "_"
+            if symbol != "_" or (i + 1) == len(melody):
+                # ensure we're dealing with notes/rests beyond the first symbol
+                if start_symbol is not None:
+                    quarter_length_duration = step_duration * step_counter # converts to the time scale of a note
+    
+                    # Handle rest
+                    # Else Handle note
+                    if start_symbol == "r":
+                        m21_event = m21.note.Rest(quarterLength=quarter_length_duration)
+                    else:
+                        m21_event = m21.note.Note(int(start_symbol), quarterLength=quarter_length_duration)
+
+                    stream.append(m21_event)
+                    
+                    # reset to the other note
+                    step_counter = 1
+                start_symbol = symbol
```

Comment:
Consider: A guard structure instead of nested if-blocks. For example 
```
if (badCondition1) {
    break
}

if (badCondition2) {
    break
}

foobar()
```

instead of if-blocks inside of if-blocks. The former makes it clearer what we don't want and easier to debug for someone who didn't create the original codebase

Reply:
This is a good suggestion, I actually want to refactor a lot of melody_generator so I am going to leave this change for the next issue involving that refactor (since technically its not an all purpose melody_generator, its only applicable to LSTMs networks), I'll reference this as a requirement for that issue